typedtree: split patterns into "value patterns" and "computation patterns"

gasche opened the pull request:
Value patterns match on a value (the result of computation), while
computation patterns handle the effects (hint hint) of
a computation. The only forms of computation patterns in OCaml today
are value patterns and exception patterns (exception p).

The sub-pattern `p` of the `lazy p` construction should be
a computation pattern, rather than a value pattern. This pull-request
makes the change, but it doesn't extend the type-checker and
pattern-matching machinery to deal with exception patterns in this
position: most of these places will `assert false` if presented with
a (lazy (exception Foo)) pattern.

Most of the changes in this PR are boilerplate -- it really is a lot
of work now to add a new syntactic category to the typed-tree
syntax. This boilerplate is fairly automatic and should be easy to
review.

There is a subtle part to the patch, though: the implementation of the
pattern type-checking. It now has to reconstruct the value/computation
distinction (absent from the parse-tree), and return values from two
different types. Instead of splitting the type-checker in several
functions (which risked code duplications), I choose to use a GADT to
have the same [type_pat] function return two different types depending
on the caller. This is the least invasive way to adapt this part of
the codebase, whose inherent complexity is so large (unfortunately)
that adding a GADT to the mix barely makes a difference.

This is the "mother PR" on top of #8959, #8967, #8968 and #8969, which are independent changes that are better reviewed independently (and make the diff here simpler and cleaner).

The implementation of the present PR is of course worth discussing, but I would also be interested in feedback on the language design aspects. I think that @trefis, @lpw25, @yallop, @maranget and @garrigue could be interested and have opinions on this matter.

gasche committed:
split patterns into "value patterns" and "computation patterns"

Value patterns match on a value (the result of computation), while
computation patterns handle the effects (hint hint) of
a computation. The only forms of computation patterns in OCaml today
are value patterns and exception patterns (exception p).

The sub-pattern `p` of the `lazy p` construction should be
a computation pattern, rather than a value pattern. This pull-request
does not make this change.

Most of the changes in this PR are boilerplate -- it really is a lot
of work now to add a new syntactic category to the typed-tree
syntax. This boilerplate is fairly automatic and should be easy to
review.

There is a subtle part to the patch, though: the implementation of the
pattern type-checking. It now has to reconstruct the value/computation
distinction (absent from the parse-tree), and return values from two
different types. Instead of splitting the type-checker in several
functions (which risked code duplications), I choose to use a GADT to
have the same [type_pat] function return two different types depending
on the caller. This is the least invasive way to adapt this part of
the codebase, whose inherent complexity is so large (unfortunately)
that adding a GADT to the mix barely makes a difference.

alainfrisch commented:
I think this PR description lacks some explicit motivation. Considering the amount of boilerplate in the PR and the fact that the only existing kind of computation pattern is supported today without an explicit category in the Typedtree, you should provide arguments of why the change will be beneficial.  Is this preparation work to allow other kinds of computation patterns, starting with `lazy (exception P)` (and later other effects)?  If so, are you convinced that the present PR will significantly simplifies such addition (or make the final code base nicer), and can you give some arguments?  Do you also have in mind other kinds of computations patterns (views, etc)?

gasche commented:
Indeed, thanks for asking.

I would say that the key motivation for the change should be: the current design is clearly not the right one. It allows to represent patterns in the typedtree that have no meaning whatsoever, such as `_ :: (exception Not_found)`. I am proposing a new design that is better, and may be the right one. Switching today has a cost, but it's clear to me that the more we wait, the higher the cost will be anyway.

Yes, it's likely that adding this distinction would make it easier to implement effect-handler, but I'm not myself planning to work on effect handlers in the short term and it was not a motivation for this PR. (That said, there is a prototype implementation of effect handlers somewhere, so those are not complete vaporware either.)

My personal motivation for the change is to get rid of the ton of `| Ppat_exception _ -> assert false` that litter (in both senses) the parts of the compiler codebase that deal with pattern-matching. As part of the refactoring work with @trefis I've been trying to design clean APIs with conversion functions between pattern representations, and having to say for each one "... and it will blow up on `Tpat_exception`" gets old fast.

See for yourself:

```shell
# search for one-liner clauses on exceptions
$ git grep "Tpat_exception.*-> *[^ ]"
ambda/matching.ml:    | Tpat_exception _ -> assert false
lambda/matching.ml:  | Tpat_exception _ -> assert false
lambda/matching.ml:  | Tpat_exception _ -> assert false
typing/parmatch.ml:  | Tpat_exception _ -> assert false
typing/parmatch.ml:| Tpat_exception _ -> assert false
typing/parmatch.ml:        | Tpat_exception _ -> assert false
typing/rec_check.ml:    | Tpat_exception _ -> false
typing/tast_iterator.ml:  | Tpat_exception p -> sub.pat sub p
typing/tast_mapper.ml:    | Tpat_exception p -> Tpat_exception (sub.pat sub p)
typing/typecore.ml:  | Tpat_array _ | Tpat_lazy _ | Tpat_exception _ -> p.pat_type
typing/typecore.ml:          | Tpat_exception _ -> res := true
typing/typedtree.ml:  | Tpat_exception p -> f p
typing/typedtree.ml:  | Tpat_exception p1 -> Tpat_exception (f p1)
typing/untypeast.ml:    | Tpat_exception p -> Ppat_exception (sub.pat sub p)
```

Out of 14 result lines, 6 are `assert false`. This makes the code harder to maintain, refactor, and reason about.

yallop commented:
> the current design is clearly not the right one. It allows to represent patterns in the typedtree that have no meaning whatsoever, such as `_ :: (exception Not_found)`

There's something rather [familiar](https://github.com/ocaml/ocaml/issues/6318#issuecomment-473025323) about this conversation.

Last time we discussed this issue, one [argument](https://github.com/ocaml/ocaml/issues/6318#issuecomment-473025360) that tipped the balance in favour of the current design was that it minimsed changes to the parse tree, avoiding disruption to ppx extensions.  Is the tooling around ppx sufficiently better now that the argument no longer carries such weight?

gasche commented:
The present proposal makes a distinction between value and computation patterns only in the typedtree, not in the parsetree phase, so ppxes are not an issue here.

Indeed, we discussed the design space back in 2014 when you (@yallop) implemented `match .. with exception`. One of the reasons against the value/computation split was "this will be more work", which isn't so important now that the work has been done in this PR, and is proposed for integration.

yallop commented:
Okay, I see.

In the original implementation [computation patterns and value patterns were split in the typed tree](https://github.com/ocaml/ocaml/commit/0f1fb19cbe48918c5d070e475c39052875623a85#diff-c91de55f58d59ae6e3535cc3eb439cd6R79), appearing as separate case lists in each `match`.

Then, when or-pattern support for exception patterns was added, [exception patterns became part of the pattern algebra](https://github.com/ocaml/ocaml/commit/369ea5f4321ff532ed1f3eb9a60feac1a79f7d7b#diff-396267e74ec9bff58486d1c43e950daaR176) giving the representation of each `match` a single list of cases.

This PR returns to the split-syntax approach, but with a grammar that supports or-patterns separately for value patterns and computation patterns.  I haven't looked at the code in detail, but that certainly seems like a better representation.

garrigue commented on typing/parmatch.ml:
```diff
@@ -24,6 +24,13 @@ open Typedtree
 
+let pure_pattern p =
+  { p with pat_desc = Tcomp_pat_return p }
+let assume_pure_pattern cp =
```
Couldn't understand the name immediately. Maybe `coerce_pure_pattern` ?

garrigue commented on typing/typedtree.mli:
```diff
@@ -35,8 +35,10 @@ type attributes = attribute list
-  { pat_desc: pattern_desc;
+type pattern = pattern_desc pattern_
+
+and 'a pattern_ =
```
`pattern_node` ?
Please reserve the final `_` for things like keywords, it looks like something you didn't want to name.

gasche replied:
There are situations (not in this PR, but in another forthcoming PR!) where we want to manipulate several distinct instances of `'a pattern_` in the same codebase, to distinguish various sub-classes of patterns in the pattern-matching analysis and compilation code. So it really helps if the name we choose for this one is short enough -- `pattern_node` would give type signatures that are really verbose. Would you prefer `'a pat` to `'a pattern_`? (Personally I like `pattern_` more, but I would be fine with `pat`).

Note: the parser has used `foo_` to mean "the auxiliary variant of `foo`" for a while now, so there is an established usage of this naming scheme outside keywords. But I don't mind using something else as long as it remains reasonably compact.

garrigue replied:
`pat` or `pattern'` are ok. For me something ending with `_` is not a name.

garrigue commented on typing/typedtree.mli:
```diff
@@ -104,10 +106,19 @@ and pattern_desc =
+and computation_pattern_desc =
+  | Tcomp_pat_exception of pattern
         (** exception P *)
+  | Tcomp_pat_return of pattern
```
Looks like monadic code... Wouldn't `Tcomp_pat_value` be what you meant?

gasche replied:
I have seen effect-handler languages where the syntax looks like

```
match e with
| return p -> ...
| Set v k -> ...
| ...
```

and it is this `return` usage that I am referring to (we are matching on a value that was returned by the expression, rather than an effectful operation that was raised). `value` would also be okay, I guess?

garrigue commented on typing/typedtree.mli:
```diff
@@ -104,10 +106,19 @@ and pattern_desc =
-  | Tpat_lazy of pattern
-        (** lazy P *)
-  | Tpat_exception of pattern
+  | Tpat_lazy of computation_pattern
```
Does this mean that we really allow exceptions under `lazy`, or is it just future work?

gasche replied:
This is left for future work, but the idea is to allow them there, yes (various code in parmatch and matching would have to be updated for this to be supported in practice). I could remove this structural change from the PR (it's not important to the reason why I made the PR), but most of the complexity in the PR comes from `type_pat` which wouldn't be simplified. I'm happy to remove if there is agreement that it is important for a first iteration.

garrigue commented on typing/typecore.ml:
```diff
@@ -1103,24 +1198,25 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
   | Ppat_constant cst ->
       let cst = constant_or_raise !env loc cst in
-      rup k {
+      k @@ rp @@ pure category @@ unif {
```
Very impressive (and successful) attempt at obfuscation.

gasche replied:
The function `rup k` is equivalent to `k ++ rp ++ unif` (I'm writing `++` for function composition in the usual mathematical order), I need to add `pure category` in the middle, only at some of the callsites. I can understand not being used to this style of chained applications, but I wouldn't call it obfuscation.

garrigue reviewed (commented):
Only superficial comments on the naming of some functions/constructors. I'll discuss the rest separately.

garrigue commented:
While I understand your goal of splitting categories semantically, the result makes me think that this is not the right timing to do that. Namely, in order to get rid of 7 `assert false`, you end up obfuscating `type_pat` on a grand scale.
Maybe the problem is that you are trying to keep too much sharing of the code. My impression is that we don't know yet exactly what we want to do with computational patterns, and this PR avoids the question by trying to preserve all possibilities, resulting in very complex code.

garrigue commented:
(I must admit that adding backtracking for GADT coverage check was also some kind of obfuscation, but at least the goal was to obtain new functionality whose correctness depended on sharing the same code.)

gasche commented:
I think that your comment on obfuscation is a bit unfair. The code as it exists today is a pain to read, and I tried hard to get the least invasive change I could do to it, to make it easy to reason about the fact that the change is behaviour-preserving. If you look only at the diff, you see some not-too-complex parts of the code turned into more-complex code, but this is in the middle of really-hairy code that is not touched at all, and that is the point.

Would you like me to first start by splitting `type_pat_aux` into several smaller functions, so that its own logic is easier to read?

alainfrisch commented:
I looked only at the changes to Typedtree, and I've to say I find it a bit unsatisfactory that:

  - We have both Tpat_or and Tcomp_pat_or. Are `Tcomp_pat_or(Tcomp_pat_return p1, Tcomp_pat_return p2)` and `Tcompt_pat_return(Tpat_or(p1, pt2))` supposed to be two equivalent representations in Typedtree?

  - Similarly, we have a `pat_extra` list (and also `pat_attributes`) both on and inside a `Tcomp_pat_return`, but they don't correspond to two different things in the syntax.  I suppose that the type-checker choose where to keep the actual info, but this looks abritrary and external tools that analyze e.g. .cmt file might be confused as to were to look the info.

Are we actually making the Typedtree a nicer representation here?

gasche commented:
@alainfrisch :

With this approach to separating value and computation patterns as two different types, we cannot do without two Or constructors. Namely, `(Foo | Bar) :: _` is really a value-or, and `(Foo | exception Bar)` is really a computation-or, we need both. One could make them the *same* constructor by adding a GADT parameter in the typedtree to distinguish between value and computation patterns, but I thought this design (which allows to share more code) would be more complex.

Having `pat_extra` in both places is actually necessary to support things like `Foo.(exception bar)` that are not represented as pattern constructors but as `pat_extra` stuff (don't ask me why, I think it's a fairly ugly design decision but this is the code as it is). My experience when writing the PR was that it actually results in very little code duplication.

gasche commented:
To put things in perspective, here is one of the cases of the function we are talking about here:

```ocaml
  | Ppat_record(lid_sp_list, closed) ->
      assert (lid_sp_list <> []);
      let opath, record_ty =
        try
          let (p0, p,_) = extract_concrete_record !env expected_ty in
          begin_def ();
          let ty = instance expected_ty in
          end_def ();
          generalize_structure ty;
          Some (p0, p, true), ty
        with Not_found -> None, newvar ()
      in
      let type_label_pat (label_lid, label, sarg) k =
        begin_def ();
        let (_, ty_arg, ty_res) = instance_label false label in
        begin try
          unify_pat_types loc !env ty_res (instance record_ty)
        with Error(_loc, _env, Pattern_type_clash(trace, _)) ->
          raise(Error(label_lid.loc, !env,
                      Label_mismatch(label_lid.txt, trace)))
        end;
        end_def ();
        generalize_structure ty_res;
        generalize_structure ty_arg;
        type_pat params sarg ty_arg (fun arg ->
          k (label_lid, label, arg))
      in
      let k' k lbl_pat_list =
        check_recordpat_labels loc lbl_pat_list closed;
        rup k {
        pat_desc = Tpat_record (lbl_pat_list, closed);
        pat_loc = loc; pat_extra=[];
        pat_type = instance record_ty;
        pat_attributes = sp.ppat_attributes;
        pat_env = !env }
      in
      let labels = params.labels in
      if params.constrs = None then
        k (wrap_disambiguate "This record pattern is expected to have"
             (mk_expected expected_ty)
             (type_label_a_list ?labels loc false !env type_label_pat opath
                lid_sp_list)
             (k' (fun x -> x)))
      else
        type_label_a_list ?labels loc false !env type_label_pat opath
          lid_sp_list (k' k)
```

I did not take the most complex case to prove a point. `Ppat_construct` is more complex than that, and `Ppat_or` is quite a bit more complex than that.

In `rup k ...`, `rup` is a continuation transformer: it takes a continuation `k` and returns an updated continuation that applies `k` after performing some effect on the mutable type graph (this is the `u` part) and growing a global mutable structure on the side (this is the `r` part). I chose to write it `k @@ rp @@ unif ...` instead (I had to split the parts to insert a coercion in the middle only in some cases). I think that while one may be less used to this style, it is not objectively more complex and, in any case, it is dwarfed by the complexity of the codebase of which we are discussing minor changes.

alainfrisch commented:
Thanks @gasche for your reply. I understand why we need to duplicate the "or-pattern" representation, the attributes and the "extra" stuff. But the fact that we need this duplication doesn't make me enthusiastic about the change. There are multiple possible (fully equivalent, but structurally different) Typedtree representations that would map back to the same Parsetree, and this doesn't look good. If this brought a clear, concrete, gain, one could live with that, but it's not yet obvious to me that encoding some kind/type information (that be reconstructed easily from the structure) into the definition the Typedtree will significantly simplify future additions.

I don't want to sound overly negative because I kind of like the direction (and extensions being hinted at), but at a first glance, the change complicates the Typedtree definition without simplifying the code. Avoiding a few `assert false` might not be worth the added complexity.

garrigue commented:
I don't think doing an even bigger rewriting at this point will help much.

> Switching today has a cost, but it's clear to me that the more we wait, the higher the cost will be anyway.
It's not clear at all to me. I've given the reason above.
One has to balance the benefit of refactoring to the cost. Here the cost is high, for a very small benefit (getting rid of 7 assert false). If we have a better idea of what we want to do with computation patterns, we may come up with a better approach, and a better justification to do it.

Another remark is that you are using GADTs. It's only on a very small scale, but you could also have used it for the `pattern_desc` type itself, which might have resulted in much smaller changes. Have you considered it? (whether it is wise to use GADTs in the core of type inference would be another question)

gasche commented:
I think that the presence of `assert false` is more of a problem than you seem to believe. It basically means that 90% of the code manipulating patterns (during pattern-matching compilation or analysis (including type-checking)) is pretending that it lives in a world where exception patterns don't exist, with the remaining 10% magically handling the cases where they can in fact appear.

When we read the documentation (that is, the type) of a function in `matching.mli` or `parmatch.mli`, how do we know whether it is partial (in particular, fails on exception patterns?) or is total? We don't. 

How can we get a good high-level understanding of which part of the code make which assumptions on the possible values? We can't.

The point of a type system is to let us reason statically about the code. Right now we are basically using dynamic typing for exception patterns: we pray that we don't meet them in the wrong situations, and we write code that just blows up if it does (but we don't know whether it can happen or it is dead code).

On the question of using GADTs in the typedtree representation themselves: that's a possibility. I think that it would make the whole thing more complex (because now every consumer of typedtree has to work with GADTs, instead of having GADTs circumscribed ot the `type_pat` function), but it may result in shorter code.

For example, instead of having two `shallow_iter_pattern_desc, shallow_iter_computation_pattern_desc` iterators, I would have a single iterator taking either a value of a first-class polymorphic type  `type pattern_desc_visitor { visit : 'a . 'a pattern_desc -> unit }`, or a simple function consuming an existential packing of the parametrized pattern type (`type any_pattern_desc = Any : 'a pattern_category * 'a pattern_desc -> any_pattern_desc`).

Would you like me to try the GADTs-in-the-typedtree approach to compare the two patches?

gasche commented:
An example of bug that arose, I believe, from our too-forgetful treatment of exception patterns is #7421, a soundness bug involving the interaction `explode` parameter for GADTs and lazy patterns. I think that the bug was there because we tried to handle the sub-pattern of a lazy pattern like any pattern, in a uniform way.

As soon as we make a type distinction between what's under `lazy ...` and value patterns, we start thinking more carefully about the code; it doesn't make sense to explode a computation pattern, and if the code had been written with this distinction in mind (as using different type forces you) I think we would have avoided the issue.

This PR leaves many partial `assume_pure` coercions in parmatch.ml; I think that taking the time to revisit them (and wonder what's the right treatment to use in each case) will reveal similar interactions that we had missed, and possibly bugs in our exhaustivity checking or other analyses.

gasche commented:
@alainfrisch on the non-canonicity of the representation.

Indeed, now `(Foo | Bar)` can be parsed as either `comp_or (return Foo) (return Bar)`, or `return (value_or Foo Bar)`. Both are valid and should result in the exact same type-checking behavior.

We could have chosen to use a different representation of computation patterns, which does not contain an unconstrained  `or` construct, but rather is just defined as `type computation_pattern_desc = { value_patterns : pattern list; exception_patterns : patter nlist }`. @yallop mentioned [above](https://github.com/ocaml/ocaml/pull/8970#issuecomment-534278863) that he used a version of this representation in an earlier version o the `match .. with exception ...` patch. This removes the non-canonicity, but it loses the nice property that the structure of the typedtree is exactly the structure of the parsetree, with extra information. Now `| Foo | exception E1 | Bar | exception E2` would result in a typed-tree that can only be shown bac to the user as `| Foo | Bar | exception E1 | exception E2`.

I think that on this specific point this PR (and the previous code) has the cake and eats it too. If you worry about canonicity, it suffices to use the `Typedtree.split_pattern` function that takes a computation pattern, and splits it maximally into value patterns and exception patterns (this corresponds to a decomposition that always prefers `value_or` to `computation_or` in the ambiguous situation of non-canonicity). In fact this function existed before (I think it comes from @trefis implementing support for `exception` under or-patterns), at the less-informative type 

```ocaml
(** Splits an or pattern into its value (left) and exception (right) parts. *)
val split_pattern : pattern -> pattern option * pattern option
```

After this PR it must take a `computation_pattern` as input.

If you want to write code that manipulates computation-patterns in a way that is guaranteed to be canonical, it suffices to call this function and then deal with the value patterns and the exception patterns. This is easy and natural, and many places of the codebase (in translcore, parmatch, etc.) already do this (and they did before my PR).

alainfrisch commented:
What about duplicated pat_extra/pat_attributes (and pat_loc, pat_env [which is mutable!], pat_type)?  That duplication would be avoided by having `Tcomp_pat_return of pattern_desc` instead of `Tcomp_pat_return of pattern`.  This would make the representation more "canonical", perhaps at the cost of complexifying the code.

I'm surprised that nobody suggested using polymorphic variant yet, to expose directly that any value patterns **is** a computed pattern (without having to wrap it in a new node that come with useless copies for pat_loc/pat_env/pat_type; and without having to have two constructors for or-patterns).

gasche commented:
Sure, I can make the `Tcomp_pat_return` argument be a pattern_desc if you would like that. (I don't think the implementation would be hard, but some parts would be a bit non-standard: in the generic iterator for example, I would reconstruct a `pattern` value to call the iterator on by using the metadata of the computation pattern with the description of the value pattern).

gasche commented:
My gut feeling on using polymorphic variants in the typedtree itself is similar to my reluctance to put GADTs there: it makes all typedtree consumers deal with them, its inferior error messages (unless you annotate very carefully) etc.

alainfrisch commented:
> Sure, I can make the `Tcomp_pat_return` argument be a pattern_desc if you would like that. (I don't think the implementation would be hard, but some parts would be a bit non-standard: in the generic iterator for example, I would reconstruct a `pattern` value to call the iterator on by using the metadata of the computation pattern with the description of the value pattern).

Assuming the general principle of the PR is adopted (I won't continue the fight against it, even though I'm still not convinced by the trade-off; but perhaps @garrigue will!), yes, everything which make the representation more canonical/self-explanatory is good. Without looking at the implementation of the type-checker, I currently don't know how I could for instance implement a tool parsing .cmt files to report patterns with a specific attribute (are attributes kept only on the Tcomp_pat_return, only under it, or on both?). Alternatively, you could document the new invariant in typedtree.mli (e.g. pat_attributes is empty on/below a Tcomp_pat_return, or both pat_attributes are necessarily identitical).

garrigue commented:
I'm not against the principle either, but wish the overhead would be smaller.
If a GADT approach works well, this could be the case, but this remains to be seen.
This would remove most changes on the code itself, but would still require additional annotations.
I'm not enthusiastic of using GADTs in typecore, but the point here would be that it would be relatively easy to remove them if there is a problem (they would just refine typing so as to remove assertions).

Actually, it is even possible to do that without GADTs, but this also requires type annotations and a few coercions:
```ocaml
type empty = {empty_ind: 'a. 'a}

type 'z t =
  | A of int
  | B of bool
  | Z of 'z

type full_t = exn t
type pure_t = empty t

let print : full_t -> string = function
  | A x -> string_of_int x
  | B x -> string_of_bool x
  | Z e -> Printexc.to_string e

let assume_pure : 'a. 'a t -> pure_t = function
  | Z _ -> assert false
  | A _ | B _ as x -> x

let as_full_t : pure_t -> full_t = function
  | Z e -> e.empty_ind
  | A _ | B _ as x -> x
```

gasche commented:
We have support for empty types now

```ocaml
type empty = |

type 'z t =
  | A of int
  | B of bool
  | Z of 'z

type full_t = exn t
type pure_t = empty t

let as_full_t : pure_t -> full_t = function
  | A _ | B _ as x -> x
```

but anyway I'm not convinced by the idea of using a more complex encoding just for the purpose of avoiding GADTs as a language feature.

I will work on improving canonicity as suggested by @alainfrisch, and give the GADT-in-the-typedtree approach a try as suggested by @garrigue, and we will see what happens.

garrigue commented:
Didn't realize that it worked that well. Actually, it looks like it is using GADT exhaustiveness to remove the need for an extra case, so we could say that GADTs have already infested the language.

garrigue commented:
It should be possible to use polymorphic variants too, but they do not mix well with the `pattern/pattern_desc` dichotomy, and you would have to modify all the code (only to add backquotes, though). In theory GADTs should be lighter here (i.e., only add a single GADT case, the other ones are still standard cases).

gasche commented:
What I have in mind for GADTs in the typedtree is to have a parameter to the `pattern_desc` type that characterizes its category (Value or Computation). So the interface would be like this, I guess (haven't worked on this approach yet):

```ocaml
and 'k pattern_desc =
  (* value patterns *)
  | Tpat_any : value pattern_desc
  | Tpat_var : Ident.t * string loc -> value pattern_desc
  | Tpat_alias : value pattern_desc * Ident.t * string loc -> value pattern_desc
  | Tpat_constant : constant -> value pattern_desc
  | Tpat_tuple : value pattern list -> value pattern_desc
  | Tpat_construct :
      Longident.t loc * constructor_description * value_pattern list ->
      value pattern_desc
  | Tpat_variant :
      label * value pattern option * row_desc ref ->
      value pattern_desc
  | Tpat_record
      (Longident.t loc * label_description * value pattern) list * closed_flag ->
      value pattern_desc
  | Tpat_array : value pattern list -> value pattern_desc
  | Tpat_or : 'k pattern * 'k pattern * row_desc option -> 'k pattern_desc
  | Tpat_lazy : computation pattern -> value pattern_desc
  (* computation patterns *)
  | Tpat_return : value pattern_desc -> computation pattern_desc
  | Tpat_exception : value pattern -> computation pattern_desc
```

alainfrisch commented:
If you go the GADT route, why not drop Tpat_return, and make sure that `(value :> computation)` (and a positive variance for pattern_desc)?  This would make the representation canonical (no duplication of satellite fields, and a single way to represent or patterns).

gasche commented:
I think that throwing subtyping into the mix is not going to end well. Those two semantical categories are distinct and it is right to have an explicit embedding of one into the other in our *intermediate* typed representation -- we are not talking about surface syntax or parse-tree here. I think it wouldn't be quite right that a syntactic choice in our surface syntax (is there an explicit keyword to distinguish the value cases of effect-handling matches?) would constrain our typed-language so much -- I'm fine with using `value pattern_desc` instead of `value pattern` to get rid of attributes ambiguity, but I think forcing us to use phantom polymorphic variants is going to far.

Maybe I'm not aware enough of the potential problems that would arise of `comp_or / value_or` ambiguity. Are there specific issues to watch for, situations where it is not satisfying to just encourage users to call `Typedtree.split_pattern` if they want a normal form? I can see the problem with the ambiguous placement of attributes (use case: anyone willing to detect a specific attribute on a specific pattern form in the typedtree), but I don't know of or-related issues.

garrigue commented:
You don't have to use subtyping: you can use only the tag `computation` in your definition, and make all `value` cases polymorphic. You will still need some explicit coercions (by which I mean, real pattern matching), but it should not be heavier than what you have now.

gasche commented:
I don't think it makes sense to allow the type `Tpat_alias : computation pattern * ident -> computation pattern`. A computation pattern matches on the effect of performing a computation, so this form of alias would name the effect itself, not the value. Similarly, I would expect a `Tpat_any : computation pattern` instance to match all evaluation effects, including raising an exception. (It is not the same as our current `_` that means "any returned value, but not an exception", and should have the type `value pattern`).

gasche commented:
@garrigue and myself discussed this PR again today. We agreed on the following TODO lists of things that would need to be changed before considering this PR again:

1. not allowing computation patterns under `Tpat_lazy` anymore for the first/simplest version of the change
2. rename `Tcomp_pat_return` into `Tcomp_pat_value`
3. try with GADTs in typedtree (but still an explicit separation with an explicit embedding of value patterns into computation patterns)

We also discussed the canonicity issues:
1. I decided during this discussion that the approach of imposing the invariant that return/value nodes have no attributes is better than trying to give them a `pattern_desc` payload. (The problem with the `pattern_desc` approach is that if you want to catch a typical value-pattern constructor with a specific attribute, you have to look at both value-patterns and return/value nodes in computation patterns. If you know that the return/value nodes never have non-empty attributes, the problem goes away.) 
2. @garrigue seemed convinced that `split_pattern` is a reasonable solution to the problem of the critical pair between {value,computation}_or and value/return.

I plan to do all of this, and then open a new PR for discussion.

alainfrisch commented:
> I decided during this discussion that the approach of imposing the invariant that return/value nodes have no attributes

Also for `pat_extra`?  What about `pat_env`?  (There are two mutable fields; will they be guaranteed to be always identical?)

gasche commented:
Yes, I think we can guarantee that `pat_extra` is also empty and `pat_env` is unchanged. (For `pat_env` this is automatic. For both `pat_attributes` and `pat_extra` this requires an explicit action on our part, because I think the default behavior of the code would rather be to always place them on the outermost node, but it's better for the user I think if they are on the innermost node, so we have ot push them inside specifically in the Tcomp_pat_value case.)

gasche commented:
I made the `Tpat_lazy` case carry value patterns again (semantically we know this is too conservative, and having computation patterns there made the change more general by forcing us to deal with computation patterns within value patterns, but removing this makes the change smaller).

I renamed `Tcomp_pat_return` into `Tcomp_pat_value`, which Jacques found more natural.

gasche commented:
I just pushed a separate commit that changes the representation to use GADTs in the parsetree, as we discussed. This does make the diff shorter in general, but I also had to add GADT annotations in more places (and first-class polymorphism for the generic iterators). Personally I like the result better than before the change, as it removes some redundancy.

I haven't done the canonicity-related changes yet, but in term of type representation everything that @garrigue asked has been done. This PR should now be ready for re-review.

garrigue commented on typing/printpat.mli:
```diff
@@ -15,8 +15,13 @@
-val pretty_line : Format.formatter -> Typedtree.pattern list -> unit
-val pretty_matrix : Format.formatter -> Typedtree.pattern list list -> unit
+val pretty_const
+    : Asttypes.constant -> string
```
Kind of picky, but this is the first time I see this format in a `mli`  (i.e. starting a line with `:`).

gasche replied:
It's used in other modules, see [asmcomp/cmmgen.mli](https://github.com/ocaml/ocaml/blob/60f5817/asmcomp/cmmgen.mli#L18-L22) for example.

garrigue replied:
It isn't an example of the style you use here, but of the _Jane Street style_.
For instance, in `misc.mli`:
```
    val is_prefix
       : equal:('a -> 'a -> bool)
      -> 'a list
      -> of_:'a list
      -> bool
```
That is, write one argument per line, the first one starting with `:`, and others with `->`.

What you did here is write the whole type on one line.
I did grep for something similar:
```
grep "^ *:[^(]*->" */*.mli */*/*.mli
ocamltest/actions_helpers.mli:  : bool -> string -> string -> out_channel -> Environments.t
middle_end/flambda/inline_and_simplify_aux.mli:  : round:int -> Inlining_cost.Threshold.t
```
I.e. only 2 instances in the whole distribution. I see no advantage to this style.

gasche replied:
I don't see any advantage to this style (I don't remember why I wrote things this way, except that I had to break lines to avoid a 80-column overrun), but I don't see any disadvantage either. Why bother discussing it at all? Surely there are more interesting things to be discussing on this work or the other on `type_pat`.

garrigue replied:
Good question. It stroke me as unusual, and it is apparently. Do you plan to turn it into a fad?
I have no deep objection, but I find it less readable (I admit it's purely instinctive, maybe influenced by punctuation rules in natural language).

garrigue reviewed (commented):
Overall, using a GADT seems to have the expected benefits: it avoids a lot of duplication, and reading the diff is much easier to follow. We may still need to convince others that it is reasonable to use them here.

trefis commented on typing/parmatch.ml:
```diff
@@ -1955,19 +1951,24 @@ and lubs ps qs = match ps,qs with
 let pressure_variants tdefs patl =
+  ignore (pressure_variants
+            (Some tdefs)
+            (List.map (fun p -> [p; omega]) patl))
```
I don't like this. I think the previous version was better: `add_row` (which you can name `mk_row` if you prefer) should produce the data we actually want. We shouldn't be mapping over its output here to add some extra information.

gasche replied:
We have a mismatch problem where the internal `pressure_variants` definition needs a fine-grained interface internally, but the external `pressure_variants` wants to expose a simpler interface and needs to bridge the gap.

In the previous version, we had an internal `pressure_variants` on just value patterns, and we exported a single `pressure_variants` that was in fact defined on computation patterns (it took care of splitting `exception` patterns away properly).

In the new version, I have two external variants: `pressure_variants` on value patterns (which didn't exist before), and then a `pressure_variants_in_computation_pattern` function on computation patterns (which is equivalent to the old external definition).

You would then expect that I now have *two* functions that have to do the ugly mapping from the external to the internal interface (wrapping `tdefs` in a `Some`, and doing Spaghetti-Monster-knows-what to those poor patterns). The trick I used is to define `pressure_variants_in_computation_pattern` in terms of the *external* function on value patterns, rather than in terms of the internal function. So now only the first external function has some ugly plumbing (it's the one you are commenting about). The second external function calls `split_pattern`, collects patterns in the expected way, and then calls the first external function on both sides.

My impression is that your analysis comes from looking at the diff in a wrong way (the change of names is confusing), rather than looking at the resulting code. I don't see the problem with this map, that does precisely the part of the previous function that was related to bridging the external-internal gap, separated from the `split_pattern`-related data churn.

trefis replied:
> We have a mismatch problem where the internal pressure_variants definition needs a fine-grained interface internally, but the external pressure_variants wants to expose a simpler interface and needs to bridge the gap.

I had indeed missed the fact that `pressure_variants` was used externally, I had assumed that only `pressure_variants_in_computation_pattern` would be.

trefis commented on typing/rec_check.ml:
```diff
@@ -1158,14 +1161,15 @@ and case : Typedtree.case -> mode -> Env.t * mode =
 (* p : m -| G
    with output m and input G
 
-   m is the mode under which the scrutinee of p is placed.
```
Why remove this comment?

trefis commented on typing/typedtree.ml:
```diff
@@ -25,8 +25,18 @@ type partial = Partial | Total
 
-type pattern =
-  { pat_desc: pattern_desc;
+type value = Value_pattern
```
That name seems a bit too generic, don't you agree?

Why don't you mint a `Pattern` submodule to put it (and all the following types) in.

gasche replied:
I thought about it but I worried that it would be out-of-style with the rest. (Contrarily to our own parmatch/matching, the structure here affects every consumer of the typedtree, which is a lot of people.) I'm willing to go that route if there is buy-in from the other reviewers.

trefis commented on typing/typedtree.ml:
```diff
@@ -117,7 +136,8 @@ and expression_desc =
       param : Ident.t;
-      body : case;
+      body : value case;
+
```
Blank line.

gasche commented:
I did a pass of changes to address the review comments. `'a pattern_` is now renamed into `'a pattern_data`.

garrigue commented:
I checked the changes. I'm fine with the code, except `printpat.mli`.

gasche commented:
I guess I would normally be pouting about the fact that it's a waste of time to discuss the formatting of a small `.mli` file, but in any case we can't merge this yet because I haven't worked on the canonicity-of-attribute-placement aspect that @alainfrisch (rightly) asked for.

gasche commented:
I just rebased this PR on top of trunk (after the changes to `type_pat`).

gasche committed:
Enforce a structure invariant in [Tpat_value p] patterns

pat_attributes and pat_extra nodes should be on the inner value
pattern, rather than on the outer computation pattern, so that a user
looking for a specific value-pattern constructor with a specific
attribute does not need to consider the Tpat_value case specifically.

(Thanks to Alain Frisch for this suggestion.)

gasche commented:
I just pushed a new commit that enforces the invariant that `Tpat_value` does not duplicate or take the attributes (or pat_extra fields) of its child. This should lift @alainfrisch's main (valid) concern with this work.

I wanted to add a testsuite entry to test this, but it turns out to be hard to do (there are no tests on attributes on the typedtree currently, I think). I sent #9078 as a separate PR that goes towards making this easier, and I would like to keep the test outside the scope of the current PR.

alainfrisch commented:
Thanks @gasche.  Does it seem feasible to turn pattern_data into a private record, with explicit builder functions, in order to really enforce the new invariant (and possibly more)? If we go into the direction of encoding structural invariants in the data structures through better layering of type definitions and GADTs, one could as well use abstraction (/privatization) to enforce even more invariants.

For instance, what about `pat_env`?  It is a mutable field, and nothing in the data structure guarantees that it will keep the same value outside and inside a value pattern. Does this property hold (which would mean that pat_env is never modified "outside" a `Tpat_value`)?

gasche commented:
Making `pat_data` abstract would be a very invasive change in the typing and pattern-matching subsystems. On the other hand, I could make the argument of `Tpat_value` a private synonym of `pattern` (this is a suggestion from @garrigue).

Note: I think that in some case expert users could wish to write typedtree fragments that do not respect this invariant (that Tpat_value attributes are empty), if they create and consume attributes on the typedtree as part of their own transformation pipeline. We are not trying to make it impossible, we are trying to make it difficult (having a default, easy way that preserves the invariant), and ensure that the typechecker itself does not do it.

gasche commented:
Re. `pat_env`: the only places where environments are mutated (in general a terrible idea) is right at the return point of `type_pat`, which sets exactly the same environment to its typed (sub)patterns; the invariant that both the parent and the child keep the same environment is preserved.

gasche commented:
@alainfrisch I pushed a last commit that makes the `Tpat_value` argument a private synonym of `pattern`. I have mixed feeling about this style which I find a bit heavy-handed, but on the other hand it did let me find one place in the compiler that I had forgotten to update (innocuously I believe, but still). 

I'll keep this commit in, assume that it makes you reasonably happy, and try to cajole @garrigue into approving the PR.

gasche committed:
typedtree: make the Tpat_value argument a private synonym

This prevents users from mistakenly constructing a Tpat_value pattern
using the natural implementation

    { pat with pat_desc = Tpat_value pat }

which breaks the attributes-placement invariant (the attributes are
duplicated with this version, instead of being placed only on the
value pattern, with empty attributes on the computation pattern).

(Suggestion from Jacques Garrigue.)

alainfrisch commented:
> Making pat_data abstract would be a very invasive change in the typing and pattern-matching subsystems.

I'm fine with the proposed solution, but I'm surprised that making pattern_data private (not abstract)  would be so invasive. I suspect most occurrences are in negative position, and the record is mostly created in Typecore.type_pat.

I'm really not asking to do that as part of this PR, but turning:
````
{
  pat_desc = d;
  pat_loc = loc; pat_extra=[];
  pat_type = instance expected_ty;
  pat_attributes = sp.ppat_attributes;
  pat_env = !env;
}
````
into
````ocaml
tpat d ~loc ~env:!env
  ~loc ~typ:(instance expected_ty)
  ~attributes:sp.ppat_attributes
````

removes some noise (optional argument for pat_extra which is often created empty; labels without disambiguating prefixes; a main "desc" unlabeled argument).

gasche commented:
There are 16 occurrences of `{ p with pat_desc = ... }` (grepping for `with pat_desc`), including 6 in matching, 2 in parmatch and one in translcore, it's a fairly common technique to iterate on a simpler pattern built from another pattern or to perform a transformation of a pattern (for example, replacing all its arguments by wildcards, or dually plugging real arguments instead of the wildcards). And I think there are other places that build patterns in different ways. In my own (for the most part still-unpublished) work with @trefis on pattern-matching, we also use this... pattern on a regular basis.

alainfrisch commented:
> Re. pat_env: the only places where environments are mutated (in general a terrible idea) is right at the return point of type_pat, which sets exactly the same environment to its typed (sub)patterns; the invariant that both the parent and the child keep the same environment is preserved.

Thanks for the info.  So after type-checking, all `pat_env` in a given case (or value_binding, or anywhere a `pattern` can appear in the Typedtree) are identical. Is this just a convenience that a copy of pat_env is kept at each sub-pattern(instead of storing it only once around the global pattern) -- and if so, would it make sense to stop doing that in order to make the invariant clearer -- or is the field used during type-checking of the pattern itself?

(Again, I'm not asking for any further change to this PR -- which I haven't reviewing in its current state -- just perhaps adding some comments of future cleanups to be done that you might have discovered while doing this work.)

gasche commented:
Now that I think of it, I wonder if we couldn't even make the field non-mutable, and use a deep `map` on the typed (sub)patterns instead of an iteration with a mutation.

I don't know enough of how `pat_env` is used to answer your question. Looking at the code, there are cases where indeed we are at the toplevel (so `pat_env` could be an argument of `Pexp_match` and stuff), but also situations where we call `Ctype.expand_head q.pat_env q.pat_type` during pattern compilation, possibly deep within a pattern. I guess it should still be possible to thread the environment argument around. Food for thought!

Would you like some extra time to consider reviewing the current state of the PR?

alainfrisch commented:
> Would you like some extra time to consider reviewing the current state of the PR?

No, if you, @garrigue and @trefis are all fine with the current state, my extra review probably wouldn't bring anything.

garrigue reviewed (approved):
Nothing more to say

gasche commented:
Thanks!

The CI is green but there is a Changes conflict. I'm planning to rebase now, and merge later today. ( @trefis, maybe you would like more time to have a look ? )
